pkg/loader/fileloader.go

```diff
@@ -18,90 +18,116 @@ package loader
-		return false
+		return nil, fmt.Errorf(
+			"no absolute path for '%s' : %v", root, err)
 	}
-	return filepath.IsAbs(fullFilePath)
+	if !fSys.IsDir(root) {
+		return nil, fmt.Errorf("absolute root '%s' must exist", root)
+	}
+	return &fileLoader{roots: []string{root}, fSys: fSys}, nil
 }
 
-// fullLocation returns some notion of a full path.
-// If location is a full file path, then ignore root. If location is relative, then
-// join the root path with the location path. Either root or location can be empty,
-// but not both. Special case for ".": Expands to current working directory.
-// Example: "/home/seans/project", "subdir/bar" -> "/home/seans/project/subdir/bar".
-func (l *fileLoader) fullLocation(root string, location string) (string, error) {
-	// First, validate the parameters
-	if len(root) == 0 && len(location) == 0 {
-		return "", fmt.Errorf("unable to calculate full location: root and location empty")
+// Returns a new Loader, which might be rooted relative to current loader.
+func (l *fileLoader) New(root string) (ifc.Loader, error) {
+	if root == "" {
+		return nil, fmt.Errorf("new root cannot be empty")
 	}
-	// Special case current directory, expanding to full file path.
-	if location == currentDir {
-		currentDir, err := os.Getwd()
-		if err != nil {
-			return "", err
-		}
-		location = currentDir
+	if isRepoUrl(root) {
+		return newGithubLoader(root, l.fSys)
 	}
-	// Assume the location is a full file path. If not, then join root with location.
-	fullLocation := location
-	if !filepath.IsAbs(location) {
-		fullLocation = filepath.Join(root, location)
+	if filepath.IsAbs(root) {
+		return l.childLoaderAt(filepath.Clean(root))
 	}
-	return fullLocation, nil
+	// Get absolute path to squeeze out "..", ".", etc. to check for cycles.
+	absRoot, err := filepath.Abs(filepath.Join(l.Root(), root))
+	if err != nil {
+		return nil, fmt.Errorf(
+			"problem joining '%s' and '%s': %v", l.Root(), root, err)
+	}
+	return l.childLoaderAt(absRoot)
 }
 
-// Load returns the bytes from reading a file at fullFilePath.
-// Implements the Loader interface.
-func (l *fileLoader) Load(location string) ([]byte, error) {
-	fullLocation, err := l.fullLocation(l.root, location)
-	if err != nil {
-		fmt.Printf("Trouble in fulllocation: %v\n", err)
+// childLoaderAt returns a new fileLoader with given root.
+func (l *fileLoader) childLoaderAt(root string) (*fileLoader, error) {
+	if !l.fSys.IsDir(root) {
+		return nil, fmt.Errorf("absolute root '%s' must exist", root)
+	}
+	if err := l.seenBefore(root); err != nil {
 		return nil, err
 	}
-	return l.fSys.ReadFile(fullLocation)
+	return &fileLoader{roots: append(l.roots, root), fSys: l.fSys}, nil
+}
+
+// seenBefore tests whether the current or any previously
+// visited root begins with the given path.
+// If so, the new path contains a previously visited root,
+// implying a cycle.
+func (l *fileLoader) seenBefore(path string) error {
+	for _, r := range l.roots {
+		if strings.HasPrefix(r, path) {
```

Comment:
I think `if r == path {` would be enough to prevent the cycle.

Reply:
I want to disallow
```
bases:
 - ..
```

Reply:
Updated the comment